File: go/pactman/environments.go
package pactman

import "fmt"

// Environment names a Pactman deployment.
//
// Endpoint hosts are declared in this file and nowhere else. Nothing else in
// this package contains a literal Pactman host.
type Environment string

// EnvironmentProduction is the Pactman production API, and the default.
const EnvironmentProduction Environment = "production"

// DefaultEnvironment is the environment used when none is configured.
const DefaultEnvironment = EnvironmentProduction

// Pactman's QA, SIT and sandbox hosts are internal and are deliberately not
// exposed here. Point at one with WithBaseURL if you have been given access.
var baseURLs = map[Environment]string{
	EnvironmentProduction: "https://entities.pactman.org",
}

// SingleCheckPath is the path of the single-check endpoint. {ein} is replaced
// with a normalized EIN.
const SingleCheckPath = "/api/entities/nonprofitcheck/v1/us/ein/{ein}"

// BulkCheckPath is the path of the bulk-check endpoint.
const BulkCheckPath = "/api/entities/nonprofitcheckbulk/v1/us/eins"

// MaxBulkEINs is the most EINs the API accepts in one bulk request. It mirrors
// the server-side limit and is declared once, here.
const MaxBulkEINs = 50

// SupportedEnvironments returns every environment name the SDK understands.
func SupportedEnvironments() []Environment {
	return []Environment{EnvironmentProduction}
}

// BaseURLForEnvironment returns the base URL for a named environment.
func BaseURLForEnvironment(environment Environment) (string, error) {
	baseURL, ok := baseURLs[environment]
	if !ok {
		return "", &ConfigurationError{msg: fmt.Sprintf(
			"unknown environment %q; supported: %s; use WithBaseURL to target a host that is not a named environment",
			string(environment), joinEnvironments(SupportedEnvironments()))}
	}

	return baseURL, nil
}

func joinEnvironments(environments []Environment) string {
	joined := ""

	for i, environment := range environments {
		if i > 0 {
			joined += ", "
		}

		joined += string(environment)
	}

	return joined
}

File: ruby/lib/pactman/nonprofit_check_plus/environments.rb
# frozen_string_literal: true

module Pactman
  # Environments, endpoint paths and the batch limit.
  module NonprofitCheckPlus
    # Named Pactman environments that are supported for public SDK use.
    #
    # Endpoint hosts are declared in this file and nowhere else. Nothing in this
    # gem should contain a literal Pactman host.
    module Environment
      # The Pactman production API.
      PRODUCTION = :production
    end

    # Base URL for each named environment.
    #
    # Pactman's QA, SIT and sandbox hosts are internal and are deliberately not
    # exposed here. Point at one with the `base_url:` option if you have been
    # given access to it.
    BASE_URLS = {
      Environment::PRODUCTION => "https://entities.pactman.org"
    }.freeze
    private_constant :BASE_URLS

    # The environment used when none is supplied.
    DEFAULT_ENVIRONMENT = Environment::PRODUCTION

    # Path of the single-check endpoint. `{ein}` is replaced with a normalized EIN.
    SINGLE_CHECK_PATH = "/api/entities/nonprofitcheck/v1/us/ein/{ein}"

    # Path of the bulk-check endpoint.
    BULK_CHECK_PATH = "/api/entities/nonprofitcheckbulk/v1/us/eins"

    # Maximum number of EINs the API accepts in one bulk request.
    #
    # This mirrors the server-side limit. It is declared once, here.
    MAX_BULK_EINS = 50

    # Returns the base URL for a named environment.
    #
    # @param environment [Symbol, String]
    # @return [String]
    # @raise [ArgumentError] for an environment this gem does not know.
    def self.base_url_for_environment(environment)
      name = environment.respond_to?(:to_sym) ? environment.to_sym : environment

      BASE_URLS.fetch(name) do
        raise ArgumentError,
              "Unknown Pactman environment #{environment.inspect}. " \
              "Supported: #{supported_environments.join(", ")}."
      end
    end

    # Every environment name the SDK understands.
    #
    # @return [Array<Symbol>]
    def self.supported_environments
      BASE_URLS.keys
    end
  end
end
